### PR TITLE
ask the system prompt from the Playground creation form

### LINKED ISSUE #617: Prevent modification of the system prompt after having started the playground's discussion
Following the work on #591 we would like to improve the behavior to be more consistent with how system prompt are used.

- The system prompt should be defined ONLY before starting the conversation.
- Once the conversation started, the system prompt is not editable anymore. 
- The user will see a message saying "Modifying the system prompt will significantly impact your entire playground. Unfortunately, you have limited options to make changes once the conversation has begun. Create a new playground and adjust the system prompt prior to initiating the interaction."

### PR DESCRIPTION
**Pull request by feloy**
### What does this PR do?

This PR moves the system prompt textarea to the creation form of the Playground.

### Screenshot / video of UI

https://github.com/projectatomic/ai-studio/assets/9973512/19cbb519-42c0-4490-a2ec-96a2ea3be1c0

### What issues does this PR fix or reference?

Fixes #617

### How to test this PR?

Create a playground with/without system prompt and check that it is displayed in the playground conversation if set.

### CODE REVIEW

File: packages/shared/src/StudioAPI.ts [hunk 1]
```diff
@@ -143,7 +143,7 @@ export abstract class StudioAPI {
    */
   abstract createSnippet(options: RequestOptions, language: string, variant: string): Promise<string>;
 
-  abstract requestCreatePlayground(name: string, model: ModelInfo): Promise<string>;
+  abstract requestCreatePlayground(name: string, model: ModelInfo, systemPrompt: string): Promise<string>;
```
Comment (axel7083) [inline packages/shared/src/StudioAPI.ts:146]:
```suggestion
  abstract requestCreatePlayground(name: string, model: ModelInfo, systemPrompt?: string): Promise<string>;
```

File: packages/frontend/src/pages/PlaygroundCreate.svelte [hunk 2]
```diff
@@ -55,7 +56,7 @@ async function submit() {
   // disable submit button
   submitted = true;
   try {
-    trackingId = await studioClient.requestCreatePlayground(playgroundName, model);
+    trackingId = await studioClient.requestCreatePlayground(playgroundName, model, systemPrompt ?? '');
```
Comment (axel7083) [inline packages/frontend/src/pages/PlaygroundCreate.svelte:59]:
```suggestion
    trackingId = await studioClient.requestCreatePlayground(playgroundName, model, systemPrompt ?? undefined);
```

File: packages/backend/src/studio-api-impl.ts [hunk 3]
```diff
@@ -74,9 +74,9 @@ export class StudioApiImpl implements StudioAPI {
       });
   }
 
-  async requestCreatePlayground(name: string, model: ModelInfo): Promise<string> {
+  async requestCreatePlayground(name: string, model: ModelInfo, systemPrompt: string): Promise<string> {
```
Comment (axel7083) [inline packages/backend/src/studio-api-impl.ts:77]:
```suggestion
  async requestCreatePlayground(name: string, model: ModelInfo, systemPrompt?: string): Promise<string> {
```

**Review by axel7083 — APPROVED** [review]
Great job ! We might want later to use something like 

```
createPlayground(options?: { system?: string, name?: string ....}) 
```

to easily add optional parameters

File: packages/frontend/src/pages/PlaygroundCreate.svelte (hunk 2 above)
Comment (feloy) [reply packages/frontend/src/pages/PlaygroundCreate.svelte:59]: That was my first attempt, but the problem I can see is that systemPrompt can be either undefined or an empty string, depending on if the user lets the textarea untouched, or edits its content and removes it. I wanted to have only one value (the empty string) to indicate the systemPrompt is empty. But as the two values are handled correctly on the other side, we can do like this if you prefer

Comment (axel7083) [reply packages/frontend/src/pages/PlaygroundCreate.svelte:59]: I think that the empty value should not be allowed, and be replaced with `undefined`

File: packages/backend/src/managers/playgroundV2Manager.ts [hunk 4]
```diff
@@ -103,6 +108,15 @@ export class PlaygroundV2Manager extends Publisher<PlaygroundV2[]> implements Di
 
     this.#conversationRegistry.createConversation(id);
 
+    if (systemPrompt !== undefined) {
```
Comment (jeffmaury) [inline packages/backend/src/managers/playgroundV2Manager.ts:111]:
```suggestion
    if (systemPrompt) {
```

**Review by jeffmaury — APPROVED** [review]
LGTM

**Review by vrothberg — APPROVED** [review]
LGTM, it looks really good

**Comment by slemeur** [comment]
I see this has been merged but the UX does not work for me
and I have a big -1 on this change

**Comment by slemeur** [comment]
We need to agreed that the experience is primarily designed for quick iterations for a developer.

*First time user*
We are asking the user to define something that is not yet known. The change it will do to the way the model interact will not be understood AND it will not be editable "on the go"

*User interacting with the model*
As a user I've defined a system prompt, but now that I'm in the playground environment, I don't see it anymore. So if I start a playground one day and get back to it another day, I'll not know what I had defined as system prompt. 
Exact same bad UX will happen when I'm creating multiple playground environments

*User iterating customizing the use of the model*
The UX proposed and involved by this change is breaking the whole purpose of the screen. 
As a user, I want to be able to adjust and iterate into the definition of how my model must interact. It's a multi-stage purposes.
- User will start without defining a system prompt
- User will adjust the settings and add a system prompt, replay the discussion to observe how it change the behavior of the model
- User will continue to adjust the system prompt and would like to be able to compare the behavior
It's going to be an iterative process, but here, because of this UX, the user is forced to each time get back to the creation of the playground environment.

**Comment by feloy** [comment]
> We need to agreed that the experience is primarily designed for quick iterations for a developer.
> 
> _First time user_ We are asking the user to define something that is not yet known. The change it will do to the way the model interact will not be understood AND it will not be editable "on the go"
> 
> _User interacting with the model_ As a user I've defined a system prompt, but now that I'm in the playground environment, I don't see it anymore. So if I start a playground one day and get back to it another day, I'll not know what I had defined as system prompt. Exact same bad UX will happen when I'm creating multiple playground environments
> 

The system prompt is visible in the chat as the first message of the discussion, with the "System Prompt" title:
 
system-propt-displayed

> _User iterating customizing the use of the model_ The UX proposed and involved by this change is breaking the whole purpose of the screen. As a user, I want to be able to adjust and iterate into the definition of how my model must interact. It's a multi-stage purposes.
> 
> * User will start without defining a system prompt
> * User will adjust the settings and add a system prompt, replay the discussion to observe how it change the behavior of the model
> * User will continue to adjust the system prompt and would like to be able to compare the behavior
>   It's going to be an iterative process, but here, because of this UX, the user is forced to each time get back to the creation of the playground environment.

**Comment by slemeur** [comment]
Ok, I did not see the system prompt getting displayed at the top. Now tested, that's better.
We'll move forward with this solution for 0.3 - but will hopefully be able to refine the experience for 1.0

**Pull request merged**